packages/flutter/lib/src/material/material_state_mixin.dart

```diff
@@ -0,0 +1,142 @@
+/// instead provide a `ValueChanged<bool>` function and allow [MaterialStateMixin]
+/// to manage the set of active [MaterialState]s, and the calling of [setState]
+/// as necessary.
+///
+/// {@tool snippet}
+/// ```dart
+/// class MyWidget extends StatefulWidget {
+///   const MyWidget(this.color, {required this.child, Key? key}) : super(key: key);
+///   final MaterialStateColor color;
+///   final Widget child;
+///   @override
+///   State createState() => MyWidgetState();
+/// }
+///
+/// class MyWidgetState extends State<MyWidget> with MaterialStateMixin<MyWidget> {
+///   @override
+///   Widget build(BuildContext context) {
+///     return InkWell(
+///       onFocusChange: updateMaterialState(MaterialState.focused),
+///       child: Container(
+///         color: widget.color.resolve(materialStates),
+///         child: widget.child,
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+@optionalTypeArgs
+mixin MaterialStateMixin<T extends StatefulWidget> on State<T> {
+  /// Managed set of active [MaterialState] values; designed to be passed to
+  /// [MaterialStateProperty.resolve] methods.
+  ///
+  /// To mutate and have [setState] called automatically for you, use
+  /// [setMaterialState], [addMaterialState], or [removeMaterialState]. Directly
+  /// mutating the set is possible, and may be necessary if you need to alter its
+  /// list without calling [setState] (and thus triggering a re-render).
+  ///
+  /// To check for a single condition, convenience getters [isPressed], [isHovered],
+  /// [isFocused], etc, are available for each [MaterialState] value.
+  Set<MaterialState> materialStates = <MaterialState>{};
+
+  /// Callback factory which accepts a [MaterialState] value and returns a
+  /// closure to mutate [materialStates] and call [setState].
+  ///
+  /// {@tool snippet}
+  /// Usage:
+  /// ```dart
+  /// class MyWidget extends StatefulWidget {
+  ///   const MyWidget({Key? key}) : super(key: key);
+  ///   @override
+  ///   State createState() => MyWidgetState();
+  /// }
+  ///
+  /// class MyWidgetState extends State<MyWidget> with MaterialStateMixin<MyWidget> {
+  ///   @override
+  ///   Widget build(BuildContext context) {
+  ///     return Container(
+  ///       color: isPressed ? Colors.black : Colors.white,
+  ///       child: InkWell(
+  ///        onHighlightChanged: updateMaterialState(MaterialState.pressed),
+  ///       ),
+  ///     );
+  ///   }
+  /// }
+  /// ```
+  /// {@end-tool}
+  ValueChanged<bool> updateMaterialState(MaterialState key, {ValueChanged<bool>? callback}) {
+    return (bool value) {
+      if (materialStates.contains(key) == value)
+        return;
+      setMaterialState(key, value);
+      callback?.call(value);
+    };
+  }
+
+  /// Mutator to mark a [MaterialState] value as either active or inactive.
+  void setMaterialState(MaterialState _state, bool isSet) {
+    return isSet ? addMaterialState(_state) : removeMaterialState(_state);
+  }
+
+
```

Comment:
nit: remove one of the blank lines here?

Reply:
➕